mlflow/sklearn/__init__.py

```diff
@@ -455,6 +460,271 @@ def load_model(model_uri):
+
+        run_id = self.get_run_id_for_model(model)
+        registered_dataset_list = self._eval_dataset_info_map[run_id][eval_dataset_name]
+
+        index = len(registered_dataset_list)
+        for i, id_i in enumerate(registered_dataset_list):
+            if eval_dataset_id == id_i:
+                index = i
+                break
+
+        if index == len(registered_dataset_list):
+            # register new eval dataset
+            registered_dataset_list.append(eval_dataset_id)
+
+        return self.gen_name_with_index(eval_dataset_name, index)
+
+    def register_prediction_result(self, model, eval_dataset_name, predict_result):
+        """
+        In `patched_predict`, register the prediction result instance with the model id and eval
+        dataset name. e.g.
+
+        ```
+        prediction_result = model_1.predict(eval_X)
+        ```
+        then we need register an item into the `pred_result_id_to_dataset_name_and_run_id` map like
+        below:
+        id(prediction_result) --> (inspected_original_var_name_of(eval_X), id(model_1))
+
+        With this map `dataset_id_to_dataset_name_and_run_id`, in following patched metric API,
+        we can query the eval dataset name and the model id via the `y_pred` argument instance.
+        """
+        value = (eval_dataset_name, self.get_run_id_for_model(model))
+        prediction_result_id = id(predict_result)
+        self._pred_result_id_to_dataset_name_and_run_id[prediction_result_id] = value
+
+        def clean_id(id_):
+            _get_autolog_training_status()._pred_result_id_to_dataset_name_and_run_id \
+                .pop(id_, None)
+
+        # When the `predict_result` object being GCed, its ID may be reused, so register a finalizer
+        # to clear the ID from the dict for preventing wrong ID mapping.
+        weakref.finalize(predict_result, clean_id, prediction_result_id)
+
+    @staticmethod
+    def gen_metric_call_command(call_fn_name, *call_pos_args, **call_kwargs):
+        """
+        Generate metric function call command string like `metric_fn(arg1, arg2, ...)`
+        Note: this method include inspecting argument variable name.
+         So should be called directly from the "patched method", to ensure it capture
+         correct argument variable name.
+        """
+
+        arg_list = []
+
+        def arg_to_str(arg):
+            if arg is not None and not np.isscalar(arg):
+                # dataset arguments
+                return _inspect_original_var_name(arg, fallback_name=f'<{arg.__class__.__name__}>')
+            elif np.isscalar(arg):
+                return str(arg)
+            else:
+                return arg.__class__.__name__
+
+        for arg in call_pos_args:
+            arg_list.append(arg_to_str(arg))
+
+        for arg_name, arg in call_kwargs.items():
+            arg_list.append(f'{arg_name}={arg_to_str(arg)}')
+
+        arg_list_str = ','.join(arg_list)
+        return f'{call_fn_name}({arg_list_str})'
```

Comment:
Just to confirm, 

```
metric_api(X, y, True, foo=True, bar=False)
```

For this function call, `gen_metric_call_command` returns a string that looks like:

```
metric_api(<DataFrame>, <ndarray>, True, foo=True, bar=False)
```

correct?

Reply:
If we can capture the variable name of X, y, it will return
`metric_api(X, y, True, foo=True, bar=False)`
otherwise return class type instead, like:
`metric_api(<DataFrame>, <ndarray>, True, foo=True, bar=False)`

Reply:
example:
![image](https://user-images.githubusercontent.com/19235986/126311401-39fb4169-8cc0-41e7-9ce1-69167120c8b2.png)

Reply:
oh, wait, seems metric info artifact the key lose the eval dataset name

Reply:
- Can we insert a space after a comma?
- Can we log arugmetns (except X and y) in the keyword form? 

```
metric_api(<DataFrame>, <ndarray>, baz=True, foo=True, bar=False)
```

Reply:
Oh, in this PR previous versions, it can print the arugmetns in the keyword form, but corey said it make code complex (we need check inspect method signature) and then I remove it.

Reply:
> Oh, in this PR previous versions, it can print the arugmetns in the keyword form, but corey said it make code complex (we need check inspect method signature) and then I remove it.

this commen?

https://github.com/mlflow/mlflow/pull/4491#discussion_r671795878

Reply:
yes. Previous I generate keyword info in method `_exactract_metric_api_call_arg_dict`

Reply:
I don't remember how your implementation looks like but args-to-kwargs conversin is quite simple like this:

```
def args_to_kwargs(args, func_signature):
    arg_names = list(func_signature.parameters.keys())
    return {key: value for key, value in zip(args, arg_names[: len(args)])}
```

Reply:
seems fine. But we need filter out the `*args` and `**kwargs` param from the signature param list.